### ISSUE TITLE
Add support for QOI (Quite OK Image) and QOA (Quite OK Audio) formats

### ISSUE THREAD

**Issue by vittorioromeo** [SFML/SFML#2969]
### Describe your feature request here

These are two formats designed by [Dominic Szablewski](http://twitter.com/phoboslab) that gained quite a lot of approval and praise in the gamedev community.

- [x] [QOI](https://qoiformat.org/)
    - PR: #3554 
- [ ] [QOA](https://qoaformat.org/)

They have a focus on simplicity (easy to implement and understand), but -- despite that -- have quite good results in terms of quality, compression speed, and decompression speed. 

The Godot Engine recently added support for them. 

I think it would be a nice addition to SFML as the format C implementations are very lightweight to add, and it shouldn't be hard at all to integrate them with our existing SFML interfaces.

### Use Cases

- Enable use of QOI assets
- Enable use of QOA assets

**Comment by Bambo-Borris** [comment]
Noticed this issue hasn't had much love or attention in those couple of days so thought I'd leave my view to help the discussion get going. Based on the research and reading I've done on the QOI & QOA formats, I'm interested in playing with them external to SFML and seeing if they offer me a benefit to load times (which audio assets currently suffer the worst load times on my current project). 

I think the fact the QOI format is available in tools such as GameMaker ties back to your mention of the traction it's gaining within game development. I hadn't seen much on this format prior to this issue so thank you @vittorioromeo for that.

I think users making games with lots textures, or a few large textures, and lots of sound effects would be interested in seeing if there is benefit in QOI/QOA formats in our SFML applications.

**Comment by eXpl0it3r** [comment]
I don't mind having those formats supported, as long as it doesn't require the use of a library.

I believe it's easier for the audio part, since we have separate file readers/writers, where as the image part currently fully relies on stb_image

As I understood it, QOI/QOA mostly just defines the specification and they offer a reference implementieren. Is the idea now that someone writes their own implementation or just copies one into SFML?

**Comment by Bambo-Borris** [comment]
> I don't mind having those formats supported, as long as it doesn't require the use of a library.
> 
From what I can tell it's a header only implementation for reading/writing/encoding/decoding for the QOI format. I haven't dug further into the QOA one here. 
> I believe it's easier for the audio part, since we have separate file readers/writers, where as the image part currently fully relies on stb_image
> 
The image case is a bit more problematic since there is no function in the QOI header to check if an images data contains the QOI magic bytes. For the load from file you could partially rely on the extension being `.qoi`, but for stream/memory loading you'd probably have to read the first 4 bytes to check it's actually a QOI file. Then if it is load QOI, if not fall back to use the stbi stuff. 

> As I understood it, QOI/QOA mostly just defines the specification and they offer a reference implementieren. Is the idea now that someone writes their own implementation or just copies one into SFML?
I think you can just utilize the header supplied by the QOI format maker. 

I checked the support list and it appears also raylib now supports this format, so I think there's some merit here for SFML to consider this, even if it's not for the v3 release.

**Comment by vittorioromeo** [comment]
> there is no function in the QOI header to check if an images data contains the QOI magic bytes.

Perhaps we could just use the `qoi_decode` function and fall back onto STB if it fails? It seems to check for the QOI magic bytes very early and return `nullptr` in case of failure:

https://github.com/phoboslab/qoi/blob/master/qoi.h#L513-L521

**Comment by vittorioromeo** [comment]
> there is no function in the QOI header to check if an images data contains the QOI magic bytes

I have sent a PR to add one, let's see if the maintainer likes the idea. https://github.com/phoboslab/qoi/pull/302

**Comment by vittorioromeo** [comment]
> > there is no function in the QOI header to check if an images data contains the QOI magic bytes
> 
> I have sent a PR to add one, let's see if the maintainer likes the idea. [phoboslab/qoi#302](https://github.com/phoboslab/qoi/pull/302)

Sadly, the maintainer decided to not merge this PR as they feel that probing for the `qoif` bytes is trivial enough to have a repeated implementation in every project consuming QOI, however they did express their excitement in adding QOI support to SFML.

**Comment by raineycat** [comment]
Hi, I was wondering if there was any interest in this issue still? I think it would be quite useful to have the option to use QOI/A files because of the speed and compression benefits.
I could draft a PR if this is still wanted.

**Comment by eXpl0it3r** [comment]
Yes, it's still something that we're interested in. Feel free to send a draft PR

**Comment by eXpl0it3r** [comment]
QOI support has been added with #3554
